pyrealm/demography/canopy.py

```diff
@@ -37,130 +191,136 @@ class Canopy:
-        """Relative radius values of stems at canopy layer heights."""
-        self.stem_crown_area: NDArray[np.float32]
-        """Stem projected crown area at canopy layer heights."""
+        self.heights: NDArray[np.float32]
+        """The vertical heights at which the canopy structure is calculated."""
+
+        self.crown_profile: CrownProfile
+        """The crown profiles of the community stems at the provided layer heights."""
         self.stem_leaf_area: NDArray[np.float32]
-        """Stem projected leaf area at canopy layer heights."""
+        """The leaf area of the crown model for each cohort by layer."""
+        self.cohort_lai: NDArray[np.float32]
+        """The leaf area index for each cohort by layer."""
+        self.cohort_f_trans: NDArray[np.float32]
+        """The fraction of light transmitted by each cohort by layer."""
+        self.cohort_f_abs: NDArray[np.float32]
+        """The fraction of light absorbed by each cohort by layer."""
+        self.f_trans: NDArray[np.float32]
+        """The fraction of light transmitted by the whole community by layer."""
+        self.f_abs: NDArray[np.float32]
+        """The fraction of light absorbed by the whole community by layer."""
+        self.transmission_profile: NDArray[np.float32]
+        """The light transmission profile for the whole community by layer."""
+        self.extinction_profile: NDArray[np.float32]
+        """The light extinction profile for the whole community by layer."""
+        self.fapar: NDArray[np.float32]
+        """The fraction of absorbed radiation for the whole community by layer."""
+        self.cohort_fapar: NDArray[np.float32]
+        """The fraction of absorbed radiation for each cohort by layer."""
+        self.stem_fapar: NDArray[np.float32]
+        """The fraction of absorbed radiation for each stem by layer."""
+        self.filled_community_area: float
+        """The area filled by crown after accounting for the crown gap fraction."""
+
+        # Check operating mode
+        if fit_ppa ^ (layer_heights is None):
+            raise ValueError("Either set fit_ppa=True or provide layer heights.")
+
+        # Set simple attributes
+        self.max_stem_height = community.stem_allometry.stem_height.max()
+        self.n_cohorts = community.number_of_cohorts
+        self.filled_community_area = community.cell_area * (
+            1 - self.canopy_gap_fraction
+        )
+
+        # Populate layer heights
+        if layer_heights is not None:
+            self.heights = layer_heights
+        else:
+            self.heights = fit_perfect_plasticity_approximation(
+                community=community,
+                canopy_gap_fraction=canopy_gap_fraction,
+                max_stem_height=self.max_stem_height,
+                solver_tolerance=solver_tolerance,
+            )
 
         self._calculate_canopy(community=community)
 
     def _calculate_canopy(self, community: Community) -> None:
```

Comment:
I would prefer to see the individual operations in `_calculate_canopy()` broken out into separate methods for testability

Reply:
I can see why.

The reason I haven't is because there are two trivial calculations and then a chain of calculations that build on each other. It _could_ be split into `calculate_leaf_area_index` and `calculate_light_transmission` but both of those would be much cleaner as methods of `Canopy` rather than standalone functions as the signatures are a bit unwieldy if we don't pass in the content of `self`.

If we do that then we have:

```python
class Canopy:
    ...
    def _calculate_canopy(self, ...):
        ...
       self._calculate_leaf_area_index()
       self._calculate_light_transmission()

    def _calculate_leaf_area_index(self, ...):
        ...

    def _calculate_light_transmission(self, ...):
        ...
```

That's still all chained together - it's not immediately clear to me that there is a huge advantage. Did you mean to break them out into standalone functions?

Reply:
@j-emberton I think I'll leave it as is because I'm going to be revisiting this section in #337.

Reply:
Ok. Just reviewing all this now.